### PR TITLE
Correctly convert units for a stacked histogram

### LINKED ISSUE #5898: Error on datetime data in stacked histogram plot
```
import datetime
import matplotlib.pyplot as plt

data = [[datetime.datetime(2015, 12, 24, 21, 13, 45), datetime.datetime(2015, 12, 30, 23, 37, 8), datetime.datetime(2015, 12, 30, 19, 43, 18), datetime.datetime(2015, 12, 30, 16, 14, 12), datetime.datetime(2015, 12, 30, 11, 32, 8), datetime.datetime(2015, 12, 29, 6, 29, 25), datetime.datetime(2015, 12, 28, 22, 54, 49), datetime.datetime(2015, 12, 28, 18, 41, 50), datetime.datetime(2015, 12, 28, 14, 25, 42), datetime.datetime(2015, 12, 28, 3, 1, 34), datetime.datetime(2015, 12, 27, 21, 10, 20), datetime.datetime(2015, 12, 27, 11, 29, 38), datetime.datetime(2015, 12, 26, 20, 34, 14), datetime.datetime(2015, 12, 26, 16, 58, 47), datetime.datetime(2015, 12, 26, 10, 54, 40), datetime.datetime(2015, 12, 25, 18, 17, 42), datetime.datetime(2015, 12, 24, 15, 44, 58), datetime.datetime(2015, 12, 25, 17, 25, 9), datetime.datetime(2015, 12, 25, 12, 33, 7), datetime.datetime(2015, 12, 30, 19, 26, 15), datetime.datetime(2015, 12, 30, 12, 46, 13), datetime.datetime(2015, 12, 30, 3, 38, 24), datetime.datetime(2015, 12, 25, 21, 11, 59), datetime.datetime(2015, 12, 25, 13, 30, 34), datetime.datetime(2015, 12, 24, 14, 6, 20)], [datetime.datetime(2015, 12, 28, 20, 59, 53), datetime.datetime(2015, 12, 27, 14, 3, 41), datetime.datetime(2015, 12, 26, 9, 37, 17)], [datetime.datetime(2015, 12, 29, 17, 18, 32)], [datetime.datetime(2015, 12, 29, 23, 15, 24)]]

plt.hist(data, stacked=True)
plt.show()
```

Using datetime objects in a stacked histogram throws following error:

```
  File "C:\Users\jadeb\Desktop\SocialFloods\tst.py", line 8, in <module>
h.hist(data, stacked=True)
  File "C:\Users\jadeb\Anaconda\envs\py34\lib\site-packages\matplotlib\__init__.py", line 1811, in inner
return func(ax, *args, **kwargs)
  File "C:\Users\jadeb\Anaconda\envs\py34\lib\site-packages\matplotlib\axes\_axes.py", line 5996, in hist
xmin = min(xmin, xi.min())
TypeError: unorderable types: datetime.datetime() < float()
```

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #5898. Hopefully the comment I've added in the code should make it clear what's going on.

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -6091,25 +6091,35 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
...
         nx = len(x)  # number of datasets
 
+        # Process unit information
+        # If doing a stacked histogram, the input is a list of datasets, so
+        # we need to do the unit conversion individually on eaach dataset
+        if stacked:
+            self._process_unit_info(xdata=x[0], kwargs=kwargs)
+            newx = []
+            for xi in x:
+                newx.append(self.convert_xunits(xi))
```
Comment (dopplershift) [inline lib/matplotlib/axes/_axes.py]: `x = [self.convert_xunits(xi) for xi in x]` ?

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -6091,25 +6091,35 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
...
         # Massage 'x' for processing.
         if input_empty:
             x = np.array([[]])
         else:
             x = cbook._reshape_2D(x, 'x')
         nx = len(x)  # number of datasets
 
+        # Process unit information
+        # If doing a stacked histogram, the input is a list of datasets, so
+        # we need to do the unit conversion individually on eaach dataset
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: each

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (dstansby) [reply lib/matplotlib/axes/_axes.py]: 🤦‍♂️ thanks...

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -6091,30 +6091,36 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
...
         else:
             x = cbook._reshape_2D(x, 'x')
         nx = len(x)  # number of datasets
 
+        # Process unit information
+        # If doing a stacked histogram, the input is a list of datasets, so
+        # we need to do the unit conversion individually on each dataset
+        if stacked:
+            self._process_unit_info(xdata=x[0], kwargs=kwargs)
+            x = [self.convert_xunits(xi) for xi in x]
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py]: I don't understand: it seems to me that the criterion for needing this step is not whether the histogram is stacked, but whether `x` is a list of datasets.

Comment (dstansby) [reply lib/matplotlib/axes/_axes.py]: I think you're right, I'll change this and add a test for multiple datasets that aren't stacked

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -6091,30 +6091,31 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
...
-        # Check whether bins or range are given explicitly.
-        binsgiven = (cbook.iterable(bins) or bin_range is not None)
-
         # basic input validation
         input_empty = np.size(x) == 0
-
         # Massage 'x' for processing.
         if input_empty:
-            x = np.array([[]])
+            x = [np.array([[]])]
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py]: Too many square brackets--I think this should be a list with one empty 1-D array, not with an empty 2-D array.

**Pull request merged**